app/precompiled.py

```diff
@@ -582,70 +586,54 @@ def rewrite_address_block(pdf):
     address = extract_address_block(pdf)
     address_regex = escape_special_characters_for_regex(address)
     address_regex = address_regex.replace("\n", r"\s*")
+
     pdf, message = redact_precompiled_letter_address_block(pdf, address_regex)
     pdf = BytesIO(pdf)
     pdf = add_address_to_precompiled_letter(pdf, address)
 
     return pdf, address, message
 
 
-def _extract_text_from_pdf(pdf, *, x, y, width, height):
+def _extract_text_from_pdf(pdf, *, x1, y1, x2, y2):
     """
     Extracts all text within a block.
-
-    pdf is a BytesIO or other file-like.
-    x, y are coordinates in mm from the top left of the page
-    width, height are lengths in mm
+    :param BytesIO pdf: pdf bytestream from which to extract
+    :param x1: horizontal location parameter for top left corner of rectangle in mm
+    :param y1: vertical location parameter for top left corner of rectangle in mm
+    :param x2: horizontal location parameter for bottom right corner of rectangle in mm
+    :param y2: vertical location parameter for bottom right corner of rectangle in mm
+    :return: multi-line address string
     """
-    ret = subprocess.run(
-        [
-            'pdftotext',
-            # -layout helps keep things on their correct lines
-            '-layout',
-            # encode output as utf-8
-            '-enc', 'UTF-8',
-            # -f and -l: only select page 1
-            '-f', '1',
-            '-l', '1',
-            # x/y coordinates in points (1/72th of an inch)
-            '-x', '{}'.format(int(x * mm)),
-            '-y', '{}'.format(int(y * mm)),
-            # width and height of area in points
-            '-W', '{}'.format(int(width * mm)),
-            '-H', '{}'.format(int(height * mm)),
-            '-',
-            '-',
-        ],
-        input=pdf.read(),
-        stdout=subprocess.PIPE
-    )
     pdf.seek(0)
-    return '\n'.join(
-        line.strip()
-        for line in ret.stdout.decode('utf-8').split('\n')
-        if line.strip()
-    )
+    doc = fitz.open("pdf", pdf)
+    page = doc[0]
+    rect = fitz.Rect(x1, y1, x2, y2)
+    words = page.getTextWords()
+    mywords = [w for w in words if fitz.Rect(w[:4]).intersects(rect)]
+    mywords.sort(key=itemgetter(3, 0))
+    group = groupby(mywords, key=itemgetter(3))
+    address = []
+    for y1, gwords in group:
+        address.append(" ".join(w[4] for w in gwords))
```

Comment:
what's going on here? Did you take this snippet from docs/stackoverflow or anything?

Can you put in some comments or links to relevant docs/examples etc to explain what's happening with all these itemgetters and array accesses?

Reply:
That's a fair point, I took the script from here:
https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/textboxtract.py

And it was referenced in the library docs here: https://pymupdf.readthedocs.io/en/latest/faq/#how-to-extract-text-from-within-a-rectangle

Should I add this information to the function's docstring?